### DISCUSSION TITLE
"mixing" api routes with fetching via swr from the dog-API so that the API-key stays hidden

### DISCUSSION THREAD

**Discussion by gregorsart** [neuefische/questions#760] [Web] [answered]
So far we used the approach from our mongoDB excursion to get the data via `api-routes` from the dog API.
We do not know how to integrate the `swr` fetch functionality inside the async handler function.
Also we do not know how to handle errors, as we are on the backend and cannot show the user visually if something went wrong.

Ideally we start fetching on the api-route page and get the Data from the dog API safely with the key and then afterwards, we expose only the fetched data from the backend to the frontend.

```
import useSWR from "swr";

const { data, error, isLoading } = useSWR("/api/user/123", fetcher);
export default async function handler(request, response) {
  await connect();
  if (request.method === "GET") {
    const services = await Service.find()
      .sort({ createdAt: -1 })
      .limit(request.query.limit);
    return response.status(200).json(services);
  }

  response.status(405).json({ message: "Method not allowed" });
}
```

We are looking forward to solutions / help / ideas!

**Comment by willywdev** [comment]
**Idea**: Fetch from the frontend and save the API Key in the .env.local
You can use Local Environment Variables with `{process.env.YOUR_NAME_OF_THE_ENV_VARIABLE}`.
And save it like this:
`YOUR_NAME_OF_THE_ENV_VARIABLE=YOUR_API_KEY`

Dont forget to hide it from git and also create a new Environment Variable in Vercel

**Reply by onemanwenttomow** [reply, marked as answer]
@willywdev This is a nice idea but it will still expose our API key on the client side. 

We will need to make a request from the client (i.e. a component) to the server (i.e. API route). The server will then make the request to the API and send the data back to the client. 

i.e. 

client (Component) -> server (API route) -> 3rd party API that needs secret -> server (API route) -> client (Component)

**Reply by onemanwenttomow** [reply]
Here is a [demo on codesandbox](https://codesandbox.io/p/sandbox/mystifying-andras-dd4qt5?file=%2Fpages%2Findex.js%3A9%2C13) to take a look at.
